### PR TITLE
lib: os: ring_buffer: Extend ring_buf_put_finish and ring_buf_get_finish

### LINKED ISSUE #27258: Ring buffer does not allow to partially put/get data
**Is your feature request related to a problem? Please describe.**
Current ring buffer implementation allow to claim the buffer for putting data (`ring_buf_put_claim`) and get pointer to the location within the buffer. When data is copied into that location `ring_buf_put_finish` is called to indicated how much data is copied. User can then claim the buffer, fill it (e.g. by passing it to the driver) and finish it. If for some reason data is not filled then `ring_buf_put_finish(buf, 0)` can be called to indicate that there is data copied.

What is not supported is ability to put portion of data because `ring_buf_put_finish` updates also internal pointer for claiming new buffer. Ring buffer fails in following scenario:

- allocate buffer for UART reception (`ring_buf_put_claim(buf, data, 8)`) and pass it to the driver
- on RX timeout receive 3 bytes (less that buffer size). Call `ring_buf_put_finish(buf, 3)`. UART is keep using the same buffer.
- allocate second buffer for UART and pass it to the driver. UART async API supports double buffering so buffer can be passed before previous one is completed. Because `ring_buf_put_finish` moved internal pointer, new buffer will overlap with previous buffer.

**Describe the solution you'd like**
In some cases it is vital that internal "claim" pointer is updated by `ring_buf_put_finish`. That is when we are actually done with the buffer. In the UART case above that would be if receiver is actually done with receiver to given buffer when timeout occurs. To support both cases i would add a `partial` flag to `ring_buf_put_finish` and `ring_buf_get_finish`. When `partial` flag is set to true then internal, "claim" pointer is not updated.

**API change**
That change would extend API with legacy behavior achieved by converting `ring_buf_put_finish(buf, n)` to  `ring_buf_put_finish(buf, false)`. Ring buffer API is not mentioned in https://docs.zephyrproject.org/latest/reference/overview.html#

### PR DESCRIPTION
**Pull request by nordic-krch**
Extended finishing calls with partial argument. Legacy behavior is
kept with partial flag set to false. Setting partial flag to true
indicates that only partial portion of claimed buffer is finished.

Fixes #27258.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

File: include/sys/ring_buffer.h [hunk 1]
```diff
@@ -315,13 +315,14 @@ uint32_t ring_buf_put_claim(struct ring_buf *buf, uint8_t **data, uint32_t size)
  * Ring buffer instance should not mix byte access and item access
  * (calls prefixed with ring_buf_item_).
  *
- * @param  buf  Address of ring buffer.
- * @param  size Number of valid bytes in the allocated buffers.
+ * @param  buf     Address of ring buffer.
+ * @param  size    Number of valid bytes in the allocated buffers.
+ * @param  partial True if more data is expected to be put in the buffer.
```
Comment (jakub-uC) [inline include/sys/ring_buffer.h:320]: Would you mind elaborating here a bit more? Just to be clear that it does not allow us to put more data than allocated memory.

File: include/sys/ring_buffer.h [hunk 2]
```diff
@@ -303,9 +303,37 @@ int ring_buf_item_get(struct ring_buf *buf, uint16_t *type, uint8_t *value,
...
+ * unclaimed after unsuccessful operation which was prematurely interrupted.
+ *
+ * If only one buffer is claimed at the time then @ref ring_buf_put_finish with
+ * option to unclaimed remaining space can be used as an alternative. However,
+ * if multiple claims occured and only partial unclaim is needed then this
+ * functiom must be used.
+ *
+ * @param  buf	Address of ring buffer.
+ * @param  size	Number of bytes to unclaim. Must be equal or less than claimed
+ *		bytes.
```
Comment (jukkar) [inline include/sys/ring_buffer.h:320]: weird indentation here

File: doc/releases/release-notes-2.4.rst [hunk 3]
```diff
@@ -72,6 +72,9 @@ API Changes
   To maintain original behaviour within user code, two argument invocations
   should be converted to pass a third argument ``FS_O_CREATE | FS_O_RDWR``.
 
+* :c:func:`ring_buf_put_finish` and :c:func:`ring_buf_get_finish` have one more
+  argument ``unclaim``. Set ``unclaim`` to true for legacy behavior.
```
Comment (jukkar) [inline doc/releases/release-notes-2.4.rst:76]: Should the unclaim argument be `false` for legacy behavior?

**Comment by nashif** [comment]
rebase please

**Comment by JonathanVanWin** [comment]
Hi, how should I use this?
I want to be able to do this:
- put_claim addr1
- put_claim addr2
- put_finish addr1
- put_claim addr3
- get_claim addr1
- get_finish addr1

addr[1-3] consecutive buffers from ring_buffer.

I tried using your pull request, and did not get it working.
The third put_claim, gives me addr2.
Does this happen because the put_finish frees from the end, and not from the start?
So the put_finish frees the addr2 and then put_claim takes gives the addr2?
What I want is to put_finish addr1, so completely cancel the put_claim, so get_claim will return 0.

**Comment by JonathanVanWin** [comment]
> Hi, how should I use this?
> I want to be able to do this:
> - put_claim addr1
> - put_claim addr2
> - put_finish addr1
> - put_claim addr3
> - get_claim addr1
> - get_finish addr1
> 
> addr[1-3] consecutive buffers from ring_buffer.
> 
> I tried using your pull request, and did not get it working.
> The third put_claim, gives me addr2.
> Does this happen because the put_finish frees from the end, and not from the start?
> So the put_finish frees the addr2 and then put_claim takes gives the addr2?
> What I want is to put_finish addr1, so completely cancel the put_claim, so get_claim will return 0. 

Fixed now by putting false in put_finish, and true in get_finish. I thought it was the other way around.
Still happy to have more clear explanation how do this las parameters work with examples: what it does for get_finish, and what it does for put_finish.
Maybe a visual explanation with the pointers of head and tail would be good.

TLDR: In my opinion, this PR is a nice addition to zephyr, useful for a lot of people, but should be explained better, otherwise people will get confused.
Maybe even add a sample, explaining different usages.
